### ISSUE TITLE
Support for temporal table queries?

### ISSUE THREAD

**Issue by groutr** [sqlalchemy/sqlalchemy#4797]
MariaDB 10.3 implemented support for temporal data tables from the SQL2011 spec.
https://mariadb.com/kb/en/library/temporal-data-tables/

I can't seem to find how these kinds of queries are supported in SQLAlchemy.  Am I missing something in the docs? I can't see how to insert the `FOR SYSTEM_TIME ALL` clause after the table in the FromClause.

**Comment by zzzeek** [comment]
This is a very extensive set of syntaxes that impact DDL in three different ways as well as SELECT.

The SELECT-level suffix can be added now either using [suffix_with](https://docs.sqlalchemy.org/en/13/core/selectable.html?highlight=suffix_with#sqlalchemy.sql.expression.Select.suffix_with) or perhaps more appropriately [with_hint](https://docs.sqlalchemy.org/en/13/core/selectable.html?highlight=with_hint#sqlalchemy.sql.expression.Select.with_hint).

the DDL part of this is entirely a big deal to implement, because it matters how these syntaxes may also be supported by other databases, or not.   if this is SQL standard stuff, then it can go straight into Core.  If not, it has to be defined in some specific mysql.* package, and even then, are these syntaxes that will be supported by MySQL? Or is this entirely just a MariaDB extension, and how likely is it that this extension will see much use given that MySQL has retained its market position pretty well.   

From there, how exactly SQLAlchemy would represent these is also tricky.   Likely, at the DDL level these would be keyword arguments:

    Column("foo", TIMESTAMP(6), mysql_generated="ALWAYS AS ROW START")
    Column("period", TIMESTAMP(6), mysql_for=func.system_time(column("start_timestamp"), column("end_timestamp"))

then how does one even reflect these things from the system catalog?  

It is always the case that I can illustrate `@compiles` recipes to make all these things happen, where custom constructs can define all these extra syntaxes.   this one is a lot of work, though.   Feel free to take a look at [@compiles](https://docs.sqlalchemy.org/en/13/core/compiler.html?highlight=compiles).  but this won't do reflection support which would be a lot of development effort. 

If this is truly only a MariaDB syntax and is not SQL standard, the way forward might be just a `@compiles` recipe that's up on the wiki.

**Comment by zzzeek** [comment]
huh, "System-versioned tables were first introduced in the SQL:2011 standard. "   and Postgresql does *not* support them?  that is weird

**Comment by groutr** [comment]
It looks like several major databases have gained support for temporal tables:
https://en.wikipedia.org/wiki/SQL:2011

**Comment by groutr** [comment]
@zzzeek, https://github.com/arkhipov/temporal_tables (Postgres extension for temporal tables).

**Comment by zzzeek** [comment]
Okey doke, SQL standard system versioned tables, here we go.     I want to try to *avoid* adding a significant number of arguments to Core because it's going to add clutter to the API.   Instead, I'd like to push as much of the verbosity into additional constructs that are simply accepted by the DDL and DQL systems.  Here is one proposal, that adds a single new construct `for_system_time()` that takes a variety of keyword arguments and is interpreted both in CREATE TABLE as one of the `*args` passed to `Table` as well as an open-ended construct that can be used with `with_hint()`:

```
from sqlalchemy import for_system_time

some_table = Table(
    "some_table",
    metadata, 
    Column('id', Integer, primary_key=True),
    Column('q', String(50), system_versioning=False)
    Column('p', String(50), system_versioning=True)
    Column("start_timestamp", TIMESTAMP(6), generated="always as row start", nullable=False),
    Column("end_timestamp", TIMESTAMP(6), generated="always as row end", nullable=False),
   system_versioning=for_system_time(period=("start_timestamp", "end_timestamp"))
)

```

also, SQL server' example:

```
   system_versioning=for_system_time(history_table="dbo.DepartmentHistory")
```

statements, note that "as_of", "between_start", "between_end", "from", "to" accept any column expression, like a literal date value, or text(), so the examples are like:

```

stmt = select([some_table]).with_hint(
   some_table,
   for_system_time(as_of=datetime.datetime(2016, 10, 9, 8, 7, 6))
)

stmt = select([some_table]).with_hint(
   some_table,
   for_system_time(between_start=text("(NOW() - INTERVAL 1 YEAR)"), between_end=text("NOW()"))
)

stmt = select([some_table]).with_hint(
   some_table,
   for_system_time(from=datetime.datetime(...), to=datetime.datetime(...))
)


```


sql server's format for comparison:
https://docs.microsoft.com/en-us/sql/relational-databases/tables/creating-a-system-versioned-temporal-table?view=sql-server-2017

if PG and everyone else is doing this then we need compiler support for as many DBs as possible and round trip tests in sqlalchemy/testing/suite/ to ensure they all work, not unlike the work done to add comment support.   These options all have to be reflected also.

**Comment by zzzeek** [comment]
well the PG syntax looks completely different, what is "CREATE TABLE employees_history (LIKE employees);"  ?     We can only implement SQL standard concepts at this level, with small variations to suit specific databases only.

**Comment by zzzeek** [comment]
Oracle's syntax also looks nothing like this.

**Comment by groutr** [comment]
@zzzeek, For automatically reflecting, the create table will have  `WITH SYSTEM VERSIONING` or `PERIOD` if either system versioning, application versioning or both are being used. It is a property of the table. The table_type in information schema will be "SYSTEM VERSIONED" if it is system versioned.

Looking at the pg extension, it looks like it emulates system versioning using triggers.  It seems the concepts of temporal tables has been "standardized" in SQL2011, but definitely not the syntax.

I'm fine with limiting the scope, my use case is pretty narrow: I only need to generate a query with Core.  I don't need to create or alter tables. Though, even the query structures are different enough across databases.

I like your DDL proposal.  For select statements, I feel like it should be attached to the table object, since system versioning is a property of the table. Something like:
```python
stmt = (select([...])
    .select_from( some_table1.for_system_time(from=datetime(...), to=datetime(...)
        .join(some_table2, ...))
    .where(...)
```

**Comment by zzzeek** [comment]
making it a method on Table is problematic for these reasons: it will be confusing vs. the DDL version of it; it won't work with ORM models; the type of object returned from this method would have to be a FROM clause which has a lot of syntactical implications which don't apply to this syntax - there is for example table.lateral(), but that because the "LATERAL" is a "selectable thing" with an alias name so you can select rows from it.  The "FOR SYSTEM" syntax here fits pretty well into the same slot as other use cases for "with_hint()" which is a very open ended consturct.  Additionally it serves my goal of not polluting FromClause with lots of miscellaneous methods that so far will see little use unless this versioned system tables thing becomes some widely used feature.

re: reflection, yes I know for MySQL we get it from CREATE TABLE but that means the regexp has to be written, tests need to be written for every possibility, the reflection also has to work with SQL Server which supports this same syntax, these are not "hard" but they are long and tedious tasks I would need help with.

**Comment by groutr** [comment]
@zzzeek I see your point. I'm trying the `.with_hint()` route.  Finally got it to produce the correct sql.  I just have to be careful not to use an aliased table. Does with_hint allow to possibility to insert the text before a possible alias?
I can probably carve out some time for this, but I would need direction on how to do those tasks.

The pg syntax `CREATE TABLE employees_history (LIKE employees);` means create a table employees_history that copies all columns, datatypes, and nullable settings from the parent table.

**Comment by zzzeek** [comment]
with_hint should be workable for aliases and dialects can also affect it so that it renders differently based on what the arguments are.

```
>>> from sqlalchemy import *
>>> t= table('t', column('q'))
>>> ta = t.alias()
>>> stmt = select([ta]).with_hint(ta, 'some hint')
>>> from sqlalchemy.dialects import mysql
>>> print(stmt.compile(dialect=mysql.dialect()))
SELECT t_1.q 
FROM t AS t_1 some hint

```

**Comment by zzzeek** [comment]
the rendering can be altered in any way using this technique:

```diff
diff --git a/lib/sqlalchemy/dialects/mysql/base.py b/lib/sqlalchemy/dialects/mysql/base.py
index 73484aea19..34803f9358 100644
--- a/lib/sqlalchemy/dialects/mysql/base.py
+++ b/lib/sqlalchemy/dialects/mysql/base.py
@@ -1287,6 +1287,17 @@ class MySQLCompiler(compiler.SQLCompiler):
     def get_from_hint_text(self, table, text):
         return text
 
+    def format_from_hint_text(self, sqltext, table, hint, iscrud):
+        hinttext = self.get_from_hint_text(table, hint)
+        if hinttext:
+            if table._is_from_container:
+
+                # optional, if isinstance(hint, ForSystemTime), etc.
+                sqltext = sqltext.replace(" AS", " %s AS" % hinttext)
+            else:
+                sqltext += " " + hinttext
+        return sqltext
+
     def visit_typeclause(self, typeclause, type_=None, **kw):
         if type_ is None:
             type_ = typeclause.type.dialect_impl(self.dialect)

```

**Comment by sebarti** [comment]
Hi zzzeek

Has this been followed up in anyway? Just curious as I would like to use temporal table in SQL Server and am stuck. If i use with_hint() it renders correctly after the FROM clause but, as expected, as a hint which escapes in SQL Server with "[...]". Is it possible to use the approach in your comment above for SQL Server as well (add a custom method to the compiler)?

Right now I'm using handwritten queries for getting AS OF snapshots but the apporach is very cumersome in terms integration into the rest of the code base.

**Comment by zzzeek** [comment]
> Hi zzzeek
> 
> Has this been followed up in anyway? 

not at all, sorry

> Just curious as I would like to use temporal table in SQL Server and am stuck. If i use with_hint() it renders correctly after the FROM clause but, as expected, as a hint which escapes in SQL Server with "[...]". Is it possible to use the approach in your comment above for SQL Server as well (add a custom method to the compiler)?

with_hint() does not apply any quoting to the text within it, "[]" sounds like a table name?  I don't have time to review my approaches above right now but the SQL Server compiler also includes the above hook.  it's just not publicly extensible, meaning you'd have to monkeypatch it.

> 
> Right now I'm using handwritten queries for getting AS OF snapshots but the apporach is very cumersome in terms integration into the rest of the code base.

I'm not finding a clear cut example of this syntax but looking at https://docs.microsoft.com/en-us/sql/relational-databases/tables/temporal-tables?view=sql-server-ver15 I'm trying:
```

from sqlalchemy import table, column, select
from sqlalchemy.dialects import mssql


t1 = table('SomeTable', column('a'), column('b'))

stmt = select(t1).with_hint(t1, "FOR SYSTEM TIME AS OF <date>")


print(stmt.compile(dialect=mssql.dialect()))
```

output:

```
SELECT [SomeTable].a, [SomeTable].b 
FROM [SomeTable] FOR SYSTEM TIME AS OF <date>

```

what is it supposed to look like?

**Comment by sebarti** [comment]
Thanks for your reply. It actually worked also for an ORM query. Now the date/datetime i want to pass comes from a web request. How can i sanitize the with_hint text argument so that i can pass dynamic timestamps? Sorry if that doesn't belong to this issue.

**Comment by zzzeek** [comment]
just show me the SQL please ?

**Comment by zzzeek** [comment]
oh in the hint?  I can think of hacky ways to get in there, which would be pretty weird.   The "hint text" should be possible that it is a compiled object, not just a string and that is missing os #6040 is added.

**Comment by zzzeek** [comment]
here's a fairly awful way that might work, try it out

```python
import datetime
from sqlalchemy import table, column, select, text, bindparam
from sqlalchemy.dialects import mssql
from sqlalchemy.sql.expression import FromClause
from sqlalchemy.ext.compiler import compiles


class HintTable(FromClause):
    def __init__(self, target, hint):
        self.target = target
        self.hint = hint
        self.name = target.name

    def __hash__(self):
        return self.target.__hash__()

    def __eq__(self, other):
        return other is self or other is self.target


@compiles(HintTable)
def _hint(element, compiler, **kw):
    return compiler.process(element.hint, **kw)


def add_system_time_hint(stmt, table_, date):
    ht = HintTable(
        table_,
        text("FOR SYSTEM TIME AS OF :date").bindparams(
            bindparam("date", date)
        ),
    )
    return stmt.with_hint(ht, "%(name)s")


if __name__ == "__main__":
    t1 = table("SomeTable", column("a"), column("b"))

    stmt = select(t1)
    stmt = add_system_time_hint(stmt, t1, datetime.datetime.today())


    compiled = stmt.compile(dialect=mssql.dialect())
    print(compiled)
    print(compiled.params)

```

**Comment by sebarti** [comment]
@zzzeek Sorry for my late reply. Thank you very much for the snippet. In general the with_hint() works to query temporal tables in SQL Server. I don't know if there are other use cases where it comes in handy for the hint text to be a compiled object but in this case i guess it really makes sense.

Another thing that i noted is the following: querying with AS OF works fine because every primary key is there at most once. But querying with BETWEEN ... AND ..., FROM ... TO ..., etc. with an ORM query (db.session.query(Mapper)) results in not all mapped entities being returned. Executing the query with the connection (db.session.execute(stmt, bind=...).fetchall()) returns all rows. I guess this is the deduplication on the sqlalchemy/python side? I mean those are special cases but is there a way to still get mapped instances from the query? Primary keys are not important in this case as history entries can't be changed anyway and are read only (history table in SQL Server has no keys at all).

**Comment by sebarti** [comment]
Ok so using Query.from_statement(stmt) also deduplicates but using the 1.4 style without calling unique() returns all rows (also if used in combination with scalars() of course). The migration [docs](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#joinedload-not-uniqued) mention that an exception will be raised if there are duplicate rows when using .joinedload() which is nice to raise the developers attention. But should the above mentioned issue not at least give a warning that there are indeed duplicate entries, even though in the case of temporal table queries it's actually wanted. Just some thoughts.

**Comment by zzzeek** [comment]
we now have requests for these in MySQL, SQL Server, and snowflake which is third party.    We should consider adding some construct around Table to support this concept semi- or fully- explicitly.

**Comment by tgross35** [comment]
Hey all,

Trying to break this down again, at least the inputs/outputs of it for mysql/mariaDB.

### Backend SQL

####  Creating table, MariaDB
Reference: https://mariadb.com/kb/en/system-versioned-tables/

There are no additional parameters needed by simply adding `WITH SYSTEM VERSIONING` to table DDL.

```SQL
CREATE TABLE t (
   x INT
) WITH SYSTEM VERSIONING;
```

This will create default "hidden" but selectable columns  `ROW_START` and `ROW_END`.  Alternatively, these can be called out explicitly:

```SQL
CREATE TABLE t(
   x INT,
   start_timestamp TIMESTAMP(6) GENERATED ALWAYS AS ROW START,
   end_timestamp TIMESTAMP(6) GENERATED ALWAYS AS ROW END,
   PERIOD FOR SYSTEM_TIME(start_timestamp, end_timestamp)
) WITH SYSTEM VERSIONING;
```

To add, `ALTER TABLE t ADD SYSTEM VERSIONING;`. To remove, `ALTER TABLE t DROP SYSTEM VERSIONING;` (see further syntax for explicit column names in the mariadb reference). Note that some special permissions will be needed (delete history). Editing versioned tables will also require temporarily changing the `system_versioning_alter_history` system flag.

Random side note - it seems as though they didn't plan for system versioning to be relevant in 2038.

#### Creating table, SQL Server

SQL Server seems slightly different, requiring `WITH (SYSTEM_VERSIONING = ON)`, and explicit declaration of `AS ROW START NOT NULL`/`AS ROW END NOT NULL`/`PERIOD FOR` columns. Unsure whether `WITH SYSTEM VERSIONING` would work, or if it will allow implicit as you can for MariaDB. Also allows specifying the history table.

More info here: https://docs.microsoft.com/en-us/sql/relational-databases/tables/creating-a-system-versioned-temporal-table?view=sql-server-2017

#### PostgreSQL

PG seems to be the last database to support these. However, it is actively being worked on, seemingly with the same MariaDB syntax from what I can tell. https://commitfest.postgresql.org/34/2316/. Probably not worth working on until it is released.

#### Accessing History

The `FOR SYSTEM_TIME` clause indicates that history will be accessed. Then any of the following:
- `AS OF arg1`: one argument, shows state at the exact time of arg1
- `BETWEEN arg1 AND arg2`: existing range syntax, all between arg1 and arg2 inclusive. [arg1,arg2]
- `FROM arg1 TO arg2`: new range syntax, include equal to arg1 but not arg2. ]arg1,arg2]
- `ALL` show all history, supposedly nonstandard but works in MariaDB and SQL Server

Arguments can be ISO strings, functions (`NOW()`), or TIMESTAMP objects

There is also the issue of insert vs. commit times, which it seems MariaDB handels by allowing versioning based on transaction IDs. Unsure how others will handle it.

SQL Server reference here: https://docs.microsoft.com/en-us/sql/relational-databases/tables/querying-data-in-a-system-versioned-temporal-table?view=sql-server-2017

### Implementation ideas

#### Table definition

We should be able to use the nonexplicit DDL most of the time, perhaps with table argument  `system_versioning=True`. This would need to create `Table.row_start` and `Table.row_end`. If these are already defined, raise an exception.

For the express definition, I would propose the following:

```python
from sqlalchemy import partition

Table(
    "table_name",
    Column("version_me_by_default", String(50)),     # Version true by default if turned on by table
    Column("dont_version_me", String(50), system_versioning=False),
    Column("start_timestamp", system_versioning="start"),
    Column("end_timestamp",  system_versioning="end"),
    system_versioning=True,
    partitioning=partition(on=partition.system_time)  # default partitioning as of mariadb 10.5.0
    # partitioning=partition(on=partition.system_time, interval=timedelta(months=1), partitions=12)
    # partitioning=partition(limit=12000)

)
```

The `system_versioning` column kwarg would imply `TIMESTAMP(6),  generated="always as row [start/end]", nullable=False`.  Then, the table system_versioning flag would search for `system_versioning="start"` and `system_versioning="end"` and do full definition if both are present, error if only one is, and do the "nonexplicit" way if not.

#### Query Object

Samples:

```python
from sqlalchemy import versioning

# SELECT * FROM t FOR SYSTEM_TIME AS OF TIMESTAMP'2016-10-09 08:07:06';
q1 = session.query(Class).versioning(versioning.asof('2016-10-09 08:07:06')).  

# SELECT * FROM t FOR SYSTEM_TIME BETWEEN (NOW() - INTERVAL 1 YEAR) AND NOW();
q2 = session.query(Class).versioning(versioning.between(DateTime.Now.AddYears(-1), DateTime.Now))    

# SELECT * FROM t FOR SYSTEM_TIME FROM '2016-01-01 00:00:00' TO '2017-01-01 00:00:00';
q3 = session.query(Class).versioning(versioning.fromto('2016-01-01 00:00:00', '2017-01-01 00:00:00'))

# SELECT * FROM t FOR SYSTEM_TIME ALL;
q4 = session.query(Class).versioning(versioning.all())
```


Anyway, I just wanted to throw some more interface ideas out there for review. I don't currently have the best picture of how to work with the sqlalchemy source; however, could anyone provide a quick list of where these changes need to take place? If so, I will try to start some developing the basic framework.

**Comment by CaselIT** [comment]
Thanks for the summary @tgross35 !

Just a note: since from 1.4 query is considered legacy, I think it's best to leave it as is, and add the new api only to `select`.

> I don't currently have the best picture of how to work with the sqlalchemy source; however, could anyone provide a quick list of where these changes need to take place?

Looking at the above:
- table object & column object updates to support these
- ddl compiler to properly render the table, allowing specific dialects to override it
- creation of the `versioning` object, and the return tokens used by the functions
- select object changes to support the new versioning api (to SelectBase ?)
- compiler support for the new options, allowing specific dialect to override it

Before starting work, I think it's best to wait for @zzzeek opinion on the above proposed syntax

**Comment by zzzeek** [comment]
> from sqlalchemy import versioning

> # SELECT * FROM t FOR SYSTEM_TIME AS OF TIMESTAMP'2016-10-09 08:07:06';
> q1 = session.query(Class).versioning(versioning.asof('2016-10-09 08:07:06')).  

why would we go through all the effort to add full blown "versioning()" methods to the first class select() / query() objects, and then still require the use of a separately imported "versioning" construct?   

some background is at https://cs.ulb.ac.be/public/_media/teaching/infoh415/tempfeaturessql2011.pdf .  it looks like this is specific to the "FOR SYSTEM TIME" extension to SQL.    at most then this is `for_system_time(<date>)` but I am still skeptical this should be first class on select().

**Comment by tgross35** [comment]
Thanks @CaselIT for the guidelines! Seems easy enough...

@zzzeek what would be preferred? If it is kept in a `select().versioning` method  then we could maybe do a single kwarg:

```python
select(Table).versioning(asof='2016-10-09 08:07:06')
select(Table).versioning(between=(DateTime.Now.AddYears(-1), DateTime.Now))
# fromto=(tup1,tup2) or all=True
```

Otherwise I think it could be added to `.where()` or `.filter()`, but I don't see a very logical/clean way to do that since it's distinct enough from the usual SQL `WHERE` statements (I possibly just don't know where in SQLALCHEMY to look). 

I was modeling the versioning import off of how `or_` is used for filtering (should have been `from sqlalchemy.versioning import asof`), but agreed nothing additional should be necessary if the method is already built in.

**Comment by zzzeek** [comment]
it should be called select().for_system_time() to match the SQL standard, then it can accept keyword arguments as you have to set up criteria.  I dont know what "DateTime.Now.AddYears" is, we dont have a date arithmetic language in SQLAlchemy at this time other than limited support addition and subtraction of intervals, so one would write that as "func.now() - datetime.timedelta(years=1)" or similar.

**Comment by CaselIT** [comment]
> I didn't realize query() was on the way out.

It isn't on the way out, in the sense that there are no plans to remove or deprecate it. It's just considered legacy and frozen wrt new functionality

**Comment by zzzeek** [comment]
OK next thing, is "FOR SYSTEM TIME" a modifier for specific tables?   in that case this would have to be specific to Table so that a query with multiple tables can target FOR SYSTEM TIME appropriately.  that is likely why I was pushing for this to be passed via with_hint(), which is a great way to associate extra details with a table in the FROM clause.     another way to do this is via a construct that works like LATERAL for example, e.g.:

    select(Table.id).select_from(Table.for_system_time(...))

or the with_hint version I suggested earlier:

    from sqlalchemy import for_system_time
    select(Table.id).with_hint(Table, for_system_time(...))

This is pretty close to what the purpose of with_hint() always was, though not exactly.   it does however offer the advantage that it applies to specific dialects, and you'd get it on Query too, it's portable, etc.:

    from sqlalchemy import for_system_time
    select(Table.id).with_hint(Table, for_system_time(...), "mariadb").with_hint(Table, some_other_thing, "mysql")

finally there could be:

   select(Table.id).for_system_time(Table, ...)

with_hint() is mostly designed very much for this kind of thing though.  that is, building on the POC at https://github.com/sqlalchemy/sqlalchemy/issues/4797#issuecomment-798451209 .

**Comment by CaselIT** [comment]
Maybe we could just have an utility method with the proper keyword args, that actually invokes `with_hint` for the user.

**Comment by zzzeek** [comment]
yes maybe, though i have a hunch things might be more flexible having a free-standing "for_system_table()" construct.   multiple APIs might cause confusion

**Comment by zzzeek** [comment]
i mean also, "with_hint()" is badly named.  "with_table_criteria()" is a better name

**Comment by zzzeek** [comment]
perhaps we can generalize with_hint() into with_table_criteria() and then introduce a hint() construct, like with_table_criteria(my_table, hint_text("+foobar"), "oracle"), something like that

**Comment by zzzeek** [comment]
bascially means, "dialect-specific stuff we want to place around where the table goes in the FROM clause"

**Comment by CaselIT** [comment]
> 
> 
> i mean also, "with_hint()" is badly named. "with_table_criteria()" is a better name

let's rename it then

**Comment by zzzeek** [comment]
well not quite a rename, with_table_crtieria() will accept objects that also give some context as to what the dialect should do with the text.    right now "hint text" is specific to a certain place that someting like an index hint goes.   a generic method would need additional context, "hey this text is for an index hint" "hey this text is for a system versioning clause".

**Comment by CaselIT** [comment]
I've added an issue, so we can think about it more.

**Comment by tgross35** [comment]
Started some very basic work on this. Here's my outline for what specifically needs to be done only for schema, nothing with select:

Table:
- sql/ddl.py: Create `AddSystemVersioning` and `DropSystemVersioning` classes
- sql/compiler.py DDLCompiler: Create `visit_add_system_versioning` and
`visit_drop_system_versioning` methods to modify table
- sql/compiler.py DDLCompiler: Add something in post_create_table to add `WITH SYSTEM VERSIONING` (as I have roughly implemented [here](https://github.com/tgross35/sqlalchemy/compare/master...tgross35:4797-versioning-dev)) or, maybe better, add `create_table_versioning` and `create_table_partitioning` methods after `post_create_table`, so there isn't some awkward call to super() in the middle of an dialect's overriding of `post_create_table`

Column:
- sql/schema.py `Column`: pop `system_versioning` kwarg in init, use to set type and add `Computed` for `GENERATED ALWAYS AS ROW START` syntax

Tests:
- test/sql/test_compiler.py DDLTest: add `test_create_system_versioning`, `test_add_system_versioning`, `test_drop_system_versioning

How do you handle options that aren't available in all dialects? E.g. SQLIte and PG (for at least a couple months) won't support this. Do you raise errors/warnings somewhere, or just override the DDLCompiler methods quietly?

I'd be happy to at least take a stab at a PR in a little while here. If somebody from SQLA gets the chance, I'd appreciate a check of my logic/plan to make sure I'm not too far off track.

**Comment by tgross35** [comment]
> the "WITH SYSTEM VERSIONING" clause is inline within the "CREATE TABLE" statement. Indexes are different because they have their own separate CREATE / DROP statement that we compile and execute separately.

Alright, I'll remove those and take a look at Alembic. I opened an issue for the concept here https://github.com/sqlalchemy/alembic/issues/956

> sqlalchemy uses zimports to handle the imports. There is a pre-commit hook that sets that up

Whoops, missed that install in the contributing doc.

**Comment by tgross35** [comment]
I unfortunately haven't had much more time to work on this (and it's sadly quite a bit above my SQLA knowledge anyway) so if anyone comes across this and wants to jump in, please do so. Otherwise, I do intend to get back to it but don't know when.

I should also point out - I don't think application versioning was mentioned at all here, but that's another useful feature that came about the same time as SV. It's also basically only in MariaDB for now, but I know PG is working on it (they seem to be looking for contributors on a SV patch). AV should probably fall into a separate issue though.

I'm surprised MySQL and others haven't implemented SV & AV yet. They've been a joy, after using them for a few months I don't think I'd ever want to go back to manually setting up triggers for history tables.

**Comment by zzzeek** [comment]
no worries, our codebase is not a quick one to jump in towards.

**Comment by tgross35** [comment]
Would a MR be accepted (or at least accepted for review) with what's needed for `CREATE TABLE`, without the `SELECT` changes? I might have a chance to work on that in not too long.

`Select` can be worked around using with_hint as described. There's really no straightforward way (that I've found) to make the DDL work though, so that bit is kind of blocking usage.

**Comment by zzzeek** [comment]
DDL is one step of it, so sure, we can accept the DDL portion

**Comment by zzzeek** [comment]
I'm for the moment moving this to 2.1, as this is a major change and I have not had the cycles to dig into it deeply, as it involves not only deep changes to the schema /compiler API it also involves database vendor functionality that is quite complicated, and is implemented by different vendors differently.  We're at a point where 2.0 has been anticipated for a long time, and as we continue to get issues / questions in the 1.4 series for APIs that have been dramatically improved, the need for 2.0 to go final is very high and I would prefer to slow down large changes coming in.

**Comment by zzzeek** [comment]
also to add, this particular feature has not been requested outside of this one issue, as well as some emails I got for someone interested in implementing for Google BigQuery.

**Comment by zzzeek** [comment]
notes left on the last gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3803/3#message-591655eb1dc21a6a6ac36048642c51671f580b14 ; for the moment I am putting this in blue sky:




this PR is obviously stalled for over a year now, and I'm certainly culpable in that.

I currently can't see merging this, for the following reasons:

*    we here propose support for the schema-level system versioning constructs but we have nothing in place for actually using these constructs, that is, we dont provide any query syntaxes. This makes me very worried that if/when we did add these syntaxes, the structure / API of the schema-level constructs would have to change. right now the feature is not very useful and seems like it would be much better as a wiki recipe using extension points for now. lots of SQLAlchemy features start out as wiki recipes where they can be vetted for awhile.

 *   the PR is extremely large due and has unrelated code refactorings and typing improvements, making it difficult to discern where the real code changes here are, as well as making it difficult to keep the branch in sync. I would want to break out all unrelated changes / typing improvements etc. into separate merges.

*    PostgreSQL / sqlite / MySQL continue to not have any support for this feature which speaks to the general relevance of the feature, additionally we have never had any users looking for this at all. obviously if this were a major SQL thing that a lot of people wanted to use, we'd be spending a lot of time on it. but there's been no request for it, no mention, I've really never seen these SQL features mentioned anywhere. MariaDB supporting this feature seems a little random, like they were trying to get a one-up on MySQL as they sometimes do (like with RETURNING, which when that was added, everyone came looking for that. that's the kind of thing we have to jump on).

*    the PR as it is, for what it actually does, allows the addition of a few simple syntaxes to a CREATE TABLE statement as well as some constraint syntaxes. This can all be done right now using @compiles recipes that we could put on the wiki. If people wanted these schema constructs right now, that's what I'd give them. Because we don't have any query-time constructs here anyway, it is premature to actually merge big schema changes. people who really need these syntaxes can could use @compiles as needed. the complexity and risk of merging this much code, plus a whole new page of docs, that is not requested at all is simply not a good idea right now.

I propose I can come up with a few expedient @compiles recipes for these syntaxes, still using the Period and SystemPeriod objects just not embedding them in the main repo yet, and putting them on the wiki under a usage recipe.
